### DISCUSSION TITLE
Inject ADF / Data Dir

### DISCUSSION THREAD

**Discussion by aZtOcKdOg** [vAmigaWeb/vAmigaWeb#170] [Ideas]
A .zip file containing amiga compatible data that can be accessed like an ADF (and saved to local storage) or appears as a DHx: or ZIP: device name for easy accessibility from any Amiga application with that functionality.

**Comment by mithrendal** [comment]
I took a look at the emulator core and what it does with exe files …
 https://github.com/vAmigaWeb/vAmigaWeb/blob/a138caf4023ea87e17f67d4b50defb76f65cb0bf/Emulator/Media/DiskFiles/EXEFile.cpp#L65

I does create a SD 880KB or HD disk 1760KB if the exe does not fit on the SD … and then it adds the exe file, also it creates a `s` directory and a startup-sequence and makes the ADF bootable to start the exe  …

We could use this as a code template to copy files from a zip file into a ADF… what do you think ? Does it make sense?

**Reply by aZtOcKdOg** [reply]
Yes that is a great approach!

I guess there is a possibility to create even larger image files using hdf later down the track (if in code add and will be a different request) so this method for ADF will be perfect! 880kb - 1.7MB is a great size for most creative tasks I can currently think of and works for any indie dev files that aren't too big! Great!

Thanks :)

**Comment by aZtOcKdOg** [comment]
I also replied directly to your response.

That is great to know about that auto boot setup feature for things! (I guess actual Amiga LHA support may come in the parent code at some point?)

**Comment by mithrendal** [comment]
<img width="633" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/43fbf22f-93e4-41ae-b3d4-e6d3c680aad8">

@aZtOcKdOg 
I have got it importing any file < 1.71MB now on my local machine ... I am not sure about the wording... can you have a look ? is it "import file ... as disk" or "on disk" or another wording?

**Reply by aZtOcKdOg** [reply]
Nice!

Are we limited by a word count / space?

(btw does it add a startup-sequence or extract the contents onto an ADF or will it simply have the imported file eg. 'frontier.lha' on the inserted ADF?)

Import >FILENAME< as disk and insert into:

or

Insert >FILENAME< as disk into:

(My preference is definitely the shorter second one as in the instance that the file is ALREADY an ADF (or disk media) then the existing dialogue popup (if using more than one drive) will read easily and is ofc self explanatory...)

**Reply by mithrendal** [reply]
>Are we limited by a word count / space?

when deployed to smaller devices ... should not be too long otherwise clipping

>(btw does it add a startup-sequence or extract the contents onto an ADF or will it simply have the imported file eg. 'frontier.lha' on the inserted ADF?)

currently it works like this... upon inserting a file it will check for its type, when it knows the filetype then everything works as normal... when it does not know the file type then it will try to write that file to a disk ... as a DD-disk (double density) when filesize is below 853kb otherwise as HD-disk (high density) 1.71MB. When file is exceeding hd capacity then a warning message is shown...

I did not create a startup sequence ... since it is not clear whether the file is executable.... that case is already handled by when files are loaded with  .exe extension.

>>My preference is definitely the shorter second one 

thank you, then we go with  

insert `todo_list.txt` as disk into:

df0: df1: ...

when the user disabled df1-df3 then it will be auto-inserted into df0 without message ... i think thats ok

**Comment by mithrendal** [comment]
the new version is online 😎 ... you can try ... though still have to write an entry to the release notes on the news page...

I changed the text to your first proposed text... because I saw no difference between importing a disk and inserting an ADF. I wanted to make the difference more obvious... here you see the version which is currently online

importing a file

<img width="814" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/58a30a77-cda8-4462-92e9-06487eb473b1">




vs inserting ADF

<img width="814" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/e11c3348-6e88-4d2f-bd4f-4d9629ba938b">

**Comment by aZtOcKdOg** [comment]
Amazing job!

Work exactly as I hoped... makes getting new files into the Amiga super fast!

Thanks (I like that wording and agree it helps differentiate seeing it live)

I made 2 'shorts' showing it do the 2 things I hoped...

https://youtube.com/shorts/jWVJRgMm6eE
(LHA File)

https://youtube.com/shorts/6fmUUvtKbOo
(Random File without extension)

o/

**Reply by mithrendal** [reply]
the 'shorts' are great but very fast and also small do you have regular youtube videos of them both ? I bet the originals were a bit slower and bigger so that one can follow more easily what you are doing... should we link them in the release notes on the news page ? what do you think?

**Reply by aZtOcKdOg** [reply]
Sounds good to me!
Yeah I can drop those same ones out at normal or 2x speed (those are 4x) or redo new ones with a more structured approach?

I can make a few showing different features! (throw some ideas at me!)

Should we make them branded / logo'd with urls and titles or just keep them raw?

(EDIT: As for size I use it in portrait usually but can do 'normal' YouTube style in landscape / pillarbox 5:4)

(EDIT2: Actually, yeah I will do new ones in landscape- stay tuned!)

**Reply by mithrendal** [reply]
> new ones with a more structured approach?

We can make a new section called “tutorials” on the vAmigaWeb homepage

![IMG_0064](https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/3d08b074-4e6a-4fc9-af16-b32a86e90d31)

And showcase different usecases maybe?

In a formfactor like this guy ones did https://www.youtube.com/watch?v=7tqqZaEwOic&t=77s

**Reply by mithrendal** [reply]
short tutorial videos that come to my mind are:
+install other system roms and how to switch between them
+import disks, modify and export disks
+ejecting drives
+keyboard mapping
+different mouse controls
+different joystick types
+game controller pairing
+have a nice cutout display area=borderless mode
+using action button shortcuts to quickly save and reload gamestate
+playing flipper game with locked action buttons
...

**Reply by mithrendal** [reply]
>This is just a quick example
(Please don't use lol)

ok then I linked your first youtube shortcut with the LHA file on [EAB](https://eab.abime.net/showthread.php?t=106882&page=4) and also on vAmigaWeb [news page](https://vamigaweb.github.io/doc/index.html). Hope that is fine for you ?... otherwise tell me if I should remove the link

**Reply by aZtOcKdOg** [reply]
Yeah that is fine for now! I will do replacements (landscape format) for the LHA shorts etc today and get you the links, after that you may want to amend those posts, but for now all good :) Thanks

**Reply by aZtOcKdOg** [reply]
Hey @mithrendal here is the new LHA video 720p25

https://youtu.be/cDXI1ioQ8zU

I have included the various steps presented in the video in the description
(EDIT: Created clickable timestamps)

Feel free to use that and amend in your posts if you approve!

Let me know if that format is fine and I will progressively make more! :) 

Feedback welcome ofc

**Reply by mithrendal** [reply]
new video is very good ... link is changed, it will become a star now ... I expect at least billions of clicks 🤤

**Reply by aZtOcKdOg** [reply]
Hey o/ Seems maybe only certain filetype extensions allow mounting of >880Kb but I can not reproduce it atm as I can not get anything over 880Kb to mount. What am I doing wrong? This was def working, I am sure

**Reply by mithrendal** [reply]
Yesterday I tested it too version `2024_2_12` and it did not work for files > 880kb on all drives

Then I checked the code and it seemed that inserting that HD disk can only be done into a drive of type HD … when inserting into a HD disk into a DD drive then it throws an exception … so I did yesterday a change… 

Before inserting a disk into a drive … do automatically check the disk type and change drive type accordingly…

This worked … I can import with current version `2024_02_21` disks into df1 to df3, df0 on the other hand still does not work with HD disks…maybe @dirkwhoffmann  designed it to be like that on purpose (because it is the built in drive which always was 3.5 DD)

I hope that I did not break something with yesterdays change… (for me it works now with HD disks, before not..)

**Reply by mithrendal** [reply]
@aZtOcKdOg I learned that AmigaOS workbench does not accept HD disks in `df0:`.   The just upgraded vAmigaWeb does bring the following message now when hd disk is accidentally inserted into df0.

_'captain_beefheart.wav.disk' is a HD disk which is not compatible with df0 - sorry, df0 only supports DD disks, please mount disk in df1 - df3, which beside DD also support HD disks._

**Reply by aZtOcKdOg** [reply]
Nice work! Well that solves that mystery- awesome, yes tested and works as described! Thanks

**Reply by mithrendal** [reply]
Todays version leaves drive selector panel open after message when user puts HD disk in df0: …. (Instead of closing it right after the message)

**Reply by aZtOcKdOg** [reply]
> short tutorial videos that come to my mind are:
> 
> +playing flipper game with locked action buttons
> 
> ...

Hey there! 

So I was trying a few approaches to a setup for 'Pinball Dreams' using Large L and R flipper buttons on either side of screen and a nice long transparent button in the bottom middle for the 'nudge / tilt'...

Where I got stumped was my various attempts at making a down arrow 'switch' type button for the spring ball launch.
On first press of button it should 'hold' ArrowDown and when that button is pressed a second time it releases ArrowDown and launches the ball...

Is there somewhere I can reference the valid commands besides the auto-suggest popups? 

I tried some JavaScript combos and almost got it working... 

Thanks for any help- I would like to know a bit more about this stuff and also continue recording some videos

**Reply by mithrendal** [reply]
@aZtOcKdOg just made a tutorial video for you. When your better video is out I will delete it. It is just for you to see how it works.

https://youtu.be/itCx69s1zSs?si=LQprfN2x06qnUzt3

The essential key here is the „lock action button“ setting, which removes the long press gesture for edit mode and makes the action buttons react on press / release correctly for gaming.



when using javascript you would do it like this to await the release of the button

```javascript
press_key('AltLeft');
await action_button_released(this_id);
release_key('AltLeft');
```

it is the same as action script `pressAltLeft=>await_ action_button_released=>releaseAltLeft`

**Reply by aZtOcKdOg** [reply]
Thanks for the reply!

Ok so I followed the video instructions and setup left and right flippers and nudge... works great! (eg hold left flipper button and left flipper stays held) 

The button for down arrow doesn't seem to work the same as in your video (or the other buttons I created and tested)

https://github.com/vAmigaWeb/vAmigaWeb/assets/64781882/08bc940f-4225-4ca6-847d-6836f52b9fc9

I created all action buttons new and was sure to enable 'lock action button' in settings. Do I need to totally remove vAmigaWeb incase there are some residual settings conflicts or something? Thanks for help!

**Reply by mithrendal** [reply]
Can you show your code for arrow down? Make a screen shot of the action button dialog

**Reply by aZtOcKdOg** [reply]
(I tried it with / without 'release')

![image](https://github.com/vAmigaWeb/vAmigaWeb/assets/64781882/52482533-aa75-4c85-8e17-d76d21646504)

**Reply by mithrendal** [reply]
Ah yes in the video I forgot ‚release‘ but it should work nevertheless because a simple ArrowDown (without release) it does a press/release and the release should then release the ArrowDown key. Strange.

Works here with:  
```
pressArrowDown=>await_action_button_released=>releaseArrowDown
```

Have you tried to cold start vAmigaWeb ?

What you can else try is:
- check if the virtual keyboard is working on keyboards Arrow down with keycap_touch_behaviour=exact setting

<img width="635" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/96d15f8d-1ac5-4c62-bc45-09a4d6a1755f">


- test the faulty action button with AmitTestKit (keyboard test) https://github.com/keirf/Amiga-Stuff/releases (for this make the button global so that it is shown when you fire up AmitTestKit.
<img width="546" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/486bbcc2-edcf-49d3-aea5-88fd95225de6">

**Reply by aZtOcKdOg** [reply]
Virtual Keyboard tested - working
keycap - exact
Fresh launch / Cold Start(?)

Testing with the app shows it recognises the input and also stays held when 'START' action button is pressed:

![image](https://github.com/vAmigaWeb/vAmigaWeb/assets/64781882/315a1eb0-4048-4450-80b5-e16d3960c078)

![image](https://github.com/vAmigaWeb/vAmigaWeb/assets/64781882/e62b666a-0a75-4611-8578-88951c99c09c)

**Reply by aZtOcKdOg** [reply]
(see keyboard tests above)

Also... I was experimenting with different code yesterday could any of it affected the function of ArrowDown and be overriding it? eg a loop or something rogue?

There are a few builds in the version section are they local? Should I delete them all?

 ![image](https://github.com/vAmigaWeb/vAmigaWeb/assets/64781882/95035439-0c02-4f8f-9e12-55d9e2b97118)

**Reply by aZtOcKdOg** [reply]
Hangon. It works!

I was just disabling/enabling settings etc and finally got to the game port 2 and switched from none -> touch joystick -> none 
...and the action button now works(?)

[EDIT: It happens on every launch. I am using Safari iPadOS 17.4.1]

Thanks man! Sorry not sure what that is?

**Reply by mithrendal** [reply]
I can reproduce the problem... 
- on fresh launch select `touch joystick` in port1
- locked action button `arrow down` gets released after some 100ms  or so...

I will open an issue and fix that later today ... it is certainly a bug.

for the pinball dreams flipper there is luckly no need for the touch joystick. disable them for now and it should work well

Thanks for finding it !

**Reply by aZtOcKdOg** [reply]
(Also a quick aside: is it possible to have an 'Opacity: 0.0' to allow fully transparent touch hotspots that don't affect the underlying bitmap pixels?)

**Reply by mithrendal** [reply]
The bug was in the game itself… pressing joystick1 fire while holding arrow down makes the ball jumping around with low energy … why did the devs of pinball dream do that ?

The bug in vAmigaWeb was simply event leaking… on touch both the action button i.e. arrow down and the fire in port1 got triggered which made the game react like it did. I tried with a physical keyboard and XBox controller and it was the same effect. What is the intention behind that feature ?

For us it means we must prevent leaking of the event meant for action button into the touch  joystick which should be easy to fix… do you remember ? we did that in the past for the key board key caps too.

Opacity should be easy too

**Reply by aZtOcKdOg** [reply]
Wow! Nice find... I wonder if the WHDLoad patched version fixed that? (I can have a look at the release notes to check!)

Thanks heaps for the help I have been playing it plenty tonight!

**Reply by aZtOcKdOg** [reply]
Thanks for the updates! Works amazingly well :D I also managed to find a HD install version for native Pinball Dreams NTSC and by using ntsc / widescreen / overscan settings in vAw we can get the game fullscreen on iPad Air 4 at butter smooth 60hz with full touch control... RAD!
[EDIT: oof github compressed the vid let me get a link]
[EDIT2: DM'd link at eab]
(Please don't share this video (well you can if you want -it's just a bit boring gameplay) I just wanted to show you the results! Will make a proper video soon!)

Thanks \o/

**Reply by aZtOcKdOg** [reply]
I can make the action button problem return (seemingly) randomly when switching between apps and/or enabling Guided Access (hide the bottom white bar) mode...

This time the only way to get the spring touch action button (ArrowDown) to work again is to switch the mouse in port 1 to 'Touch Joystick' etc...

**Comment by mithrendal** [comment]
>There a few builds in the version section are they local? Should I delete them all?

yes lets do some house keeping and delete them... 😀 but should not affect the problem

with cold start I mean swipe the app out of memory in iPadOS  so that all javascript tests which you have done before are cleared... Maybe there are too many releaseEvents for that ArrowDown key in the keyboard buffer hidden ?

Also check how many entries you see in the list of action buttons ... maybe in pinball dreams you have some hidden actionbutton still active which interferes?  In AmitTestKit all looks good.

**Comment by aZtOcKdOg** [comment]
Has vAmiga core had any changes regarding being able to mount >1.71Mb files and archives in a free HDD (DHxx) slot that you recall?

Just doing some iPad Amiga'ing with some file management / os install and mounting large files would be suppper handy!

AMMIIIIGGGAAAAAAA!

**Reply by mithrendal** [reply]
I think there is no API for that. See also here where another user asked for it and dirks reply

https://github.com/dirkwhoffmann/vAmiga/issues/904#issuecomment-2885864789


<img width="766" alt="image" src="https://github.com/user-attachments/assets/5e73084b-5a38-4609-8a2d-7e0648a5cb61" />
